src/worker/vm/upgrades/export-events.ts

```diff
@@ -0,0 +1,113 @@
+    jobs: {
+        exportEventsWithRetry: ExportEventsJobPayload
+    }
+}>
+
+interface ExportEventsJobPayload extends Record<string, any> {
+    batch: PluginEvent[]
+    batchId: number
+    retriesPerformedSoFar: number
+}
+
+/*
+ This function injects code into a plugin VM that has exported `exportEvents`:
+ - add the global/config/jobs specified in the `ExportEventsUpgrade` type above.
+ - patch `onEvent` with code to add the event to a buffer.
+ */
+export function upgradeExportEvents(
+    response: PluginConfigVMInternalResponse<PluginMeta<ExportEventsUpgrade>>,
+    console: ConsoleExtension
+): void {
+    const { methods, tasks, meta } = response
+
+    if (!methods.exportEvents) {
+        return
+    }
+
+    const nanToNull = (nr: any): null | number => (isNaN(parseInt(nr)) ? null : parseInt(nr))
+    const uploadBytes = Math.max(
+        1,
+        Math.min(nanToNull(meta.config.exportEventsBufferBytes) ?? 1024 * 1024, 100 * 1024 * 1024)
+    )
+    const uploadSeconds = Math.max(1, Math.min(nanToNull(meta.config.exportEventsBufferSeconds) ?? 10, 600))
+
+    meta.global.exportEventsToIgnore = new Set(
+        meta.config.exportEventsToIgnore
+            ? meta.config.exportEventsToIgnore.split(',').map((event: string) => event.trim())
+            : null
+    )
+
+    meta.global.exportEventsBuffer = createBuffer({
+        limit: uploadBytes,
+        timeoutSeconds: uploadSeconds,
+        onFlush: async (batch) => {
+            const jobPayload = {
+                batch,
+                batchId: Math.floor(Math.random() * 1000000),
+                retriesPerformedSoFar: 0,
+            }
+            const firstThroughQueue = false // TODO: might make sense sometimes? e.g. when we are processing too many tasks already?
+            if (firstThroughQueue) {
+                await meta.jobs.exportEventsWithRetry(jobPayload).runNow()
+            } else {
+                await meta.global.exportEventsWithRetry(jobPayload, meta)
+            }
+        },
+    })
+
+    meta.global.exportEventsWithRetry = async (
+        payload: ExportEventsJobPayload,
+        meta: PluginMeta<ExportEventsUpgrade>
+    ) => {
+        try {
+            await methods.exportEvents(payload.batch)
+        } catch (err) {
+            if (err instanceof RetryError) {
+                if (payload.retriesPerformedSoFar < 15) {
+                    const nextRetrySeconds = 2 ** payload.retriesPerformedSoFar * 3
+                    console.log(`Enqueued batch ${payload.batchId} for retry in ${Math.round(nextRetrySeconds)}s`)
+
+                    await meta.jobs
+                        .exportEventsWithRetry({ ...payload, retriesPerformedSoFar: payload.retriesPerformedSoFar + 1 })
+                        .runIn(nextRetrySeconds, 'seconds')
+                } else {
+                    console.log(
+                        `Dropped batch ${payload.batchId} after retrying ${payload.retriesPerformedSoFar} times`
+                    )
```

Comment:
Ok, agreed. Sent to system. I originally implemented this as a way to give the user some status about exports, mainly to see that everything is working well. However yes, this is more system level information.

Because of https://github.com/PostHog/plugin-server/issues/424 I also fear having too many console.logs, so I'm happy to remove them for stability as well.